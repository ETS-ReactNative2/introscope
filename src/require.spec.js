import { resolve } from 'path';

const introscope = path => require(path + '?introscope.js');

describe('require.introscope', () => {
    it('does not cache when require() goes first', () => {
        // in each require order test filenames should be uniqe
        const file = './require1';

        // cache the original transformation first
        const firstRequire = require(file);
        expect(typeof firstRequire).toBe('object');
        expect(firstRequire.publicVariable).toBe('publicValue');
        expect(firstRequire.privateVariable).toBe(undefined);

        // require with introscope enabled
        const firstIntroscope = introscope(file);
        expect(typeof firstIntroscope).toBe('function');
        const scope1 = firstIntroscope();
        expect(scope1.publicVariable).toBe('publicValue');
        expect(scope1.privateVariable).toBe('privateValue');

        // get again the original transformation
        const secondRequire = require(file);
        expect(typeof secondRequire).toBe('object');
        expect(secondRequire.publicVariable).toBe('publicValue');
        expect(secondRequire.privateVariable).toBe(undefined);
    });

    it('does not cache when require.introscope() goes first', () => {
        // in each require order test filenames should be uniqe
        const file = './require2';

        // require with introscope enabled first
        const firstIntroscope = introscope(file);
        expect(typeof firstIntroscope).toBe('function');
        const scope1 = firstIntroscope();
        expect(scope1.publicVariable).toBe('publicValue');
        expect(scope1.privateVariable).toBe('privateValue');

        // try getting original transformation second
        const firstRequire = require(file);
        expect(typeof firstRequire).toBe('object');
        expect(firstRequire.publicVariable).toBe('publicValue');
        expect(firstRequire.privateVariable).toBe(undefined);

        // require again with introscope enabled
        const secondIntroscope = introscope(file);
        expect(typeof secondIntroscope).toBe('function');
        const scope2 = secondIntroscope();
        expect(scope2.publicVariable).toBe('publicValue');
        expect(scope2.privateVariable).toBe('privateValue');
    });
});

describe('source maps', () => {
    it('stack trace is correct', () => {
        expect.assertions(2);
        const errorScope = introscope('./error.js');
        expect(() => {
            try {
                errorScope().throws();
            } catch (err) {
                const root = resolve(__dirname, '../../');
                expect(
                    err.stack
                        .split('/Users/peter/w/babel-module-function')
                        .join('__dirname')
                ).toMatchSnapshot();
                throw err;
            }
        }).toThrow();
    });
});
