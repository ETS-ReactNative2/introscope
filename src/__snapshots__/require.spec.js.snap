// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source maps stack trace is correct 1`] = `
"TypeError: Cannot read property 'nonExisting' of undefined
    at Object._scope.throws (__dirname/src/error.js?introscope.js:5:29)
    at __dirname/src/require.spec.js:62:30
    at Object.<anonymous> (__dirname/node_modules/expect/build/to_throw_matchers.js:43:7)
    at Object.throwingMatcher [as toThrow] (__dirname/node_modules/expect/build/index.js:214:24)
    at Object.<anonymous> (__dirname/src/require.spec.js:72:12)
    at Object.asyncFn (__dirname/node_modules/jest-jasmine2/build/jasmine_async.js:82:37)
    at resolve (__dirname/node_modules/jest-jasmine2/build/queue_runner.js:52:12)
    at new Promise (<anonymous>)
    at mapper (__dirname/node_modules/jest-jasmine2/build/queue_runner.js:39:19)
    at promise.then (__dirname/node_modules/jest-jasmine2/build/queue_runner.js:73:82)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)"
`;
